src/modules/crypto/crypto_adaptors/common.rs

```diff
@@ -0,0 +1,339 @@
+                cipher = Cipher::aes_128_gcm();
+                aead = true;
+            },
+            (AESKeySize::AES192, CipherMode::GCM) => {
+                cipher = Cipher::aes_192_gcm();
+                aead = true;
+            },
+            (AESKeySize::AES256, CipherMode::GCM) => {
+                cipher = Cipher::aes_256_gcm();
+                aead = true;
+            },
+            _ => return Err(RvError::ErrCryptoCipherOPNotSupported),
+        }
+
+        if aead == false {
+            let ciphertext = encrypt(
+                cipher,
+                &$aes.key,
+                Some(&$aes.iv),
+                $plaintext).unwrap();
+            return Ok(ciphertext.to_vec());
+        } else {
+            // aes_xxx_gcm's tag is at most 16-bytes long.
+            let tag: &mut [u8] = &mut [0; 16];
+            let ciphertext = encrypt_aead(
+                cipher,
+                &$aes.key,
+                Some(&$aes.iv),
+                &$aes.aad.clone().unwrap(),
+                $plaintext,
+                tag
+            ).unwrap();
+            $aes.tag = Some(tag.to_vec());
+            return Ok(ciphertext.to_vec());
+        }
+    }
+}
+
+macro_rules! common_aes_decrypt {
+    ($aes: expr, $ciphertext: expr) => {
+        match $aes.alg {
+            (AESKeySize::AES128, CipherMode::CBC) => {
+                let plaintext = decrypt(
+                    Cipher::aes_128_cbc(),
+                    &$aes.key,
+                    Some(&$aes.iv),
+                    $ciphertext).unwrap();
+                return Ok(plaintext.to_vec());
+            }
+            (AESKeySize::AES128, CipherMode::GCM) => {
+                // aes_128_gcm's tag is 16-bytes long.
+                let plaintext = decrypt_aead(
+                    Cipher::aes_128_gcm(),
+                    &$aes.key,
+                    Some(&$aes.iv),
+                    &$aes.aad.clone().unwrap(),
+                    $ciphertext,
+                    &$aes.tag.clone().unwrap()
+                ).unwrap();
+                return Ok(plaintext.to_vec());
+            }
+            _ => { return Err(RvError::ErrCryptoCipherOPNotSupported); }
+        }
+    }
+}
+
+macro_rules! common_aes_encrypt_update {
+    ($aes: expr, $plaintext: expr, $ciphertext: expr) => {
+        let cipher;
+
+        match $aes.alg {
+            (AESKeySize::AES128, CipherMode::CBC) => {
+                cipher = Cipher::aes_128_cbc();
+            }
+            (AESKeySize::AES128, CipherMode::GCM) => {
+                cipher = Cipher::aes_128_gcm();
+            }
+            _ => { return Err(RvError::ErrCryptoCipherOPNotSupported); }
+        }
+
+        if let None = $aes.ctx {
+            // init adaptor ctx if it's not inited.
+            let encrypter = Crypter::new(
+                cipher,
+                Mode::Encrypt,
+                &$aes.key,
+                Some(&$aes.iv)
+            ).unwrap();
```

Comment:
Same as above.

Reply:
This panic is intentionally. If a new Crypter from OpenSSL is not successfully inittialized, then there must be something very badly wrong. Even if the process returns to upper-level, caller needs to panic as well.

Reply:
It's unlike the `encrypt` or `encrypt_update` functions that allow no panic revocated.